File: codeforces/1000/1111c.py
def dp(l,r,A,B,sufix):
    if r-l<2:
        if sufix[l]-sufix[l-1]:
            return B
        return A
    ans = A
    if sufix[r-1]-sufix[l-1]:
        ans = (r-l)*B*(sufix[r-1]-sufix[l-1])
    mid = int((r+l)/2)
    res = dp(l,mid,A,B,sufix) + dp(mid,r,A,B,sufix)
    return min(ans,res)
def solve():
    # [n,k,A,B]=[int(x) for x in input().split()]
    n,k,A,B=map(int,input().split())
    # v = [int(x) for x in input().split()]
    v=list(map(int,input().split()))
    N=(1<<n)
    p = [0]*N
    for i in range(k):
        p[v[i]-1]+=1 # ojo +=
    sufix = [0]*(N+1)
    for i in range(N):
        sufix[i+1] = sufix[i]+p[i]
    return dp(1,N+1,A,B,sufix)

def init():
    T=1
    # T=int(input())
    while T:
        print(solve())
        T-=1
init()

File: codeforces/1000/1111c.cpp
#include<bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef pair<ll, ll> ii;
typedef vector<ll> vl;
typedef vector<ii> vll;

#define N6 1000006
#define N5 100005

ll n, k, A, B, T = 1;
vector<ll>p;
ll dp(ll l,ll r,ll A,ll B){
    ll i = lower_bound(p.begin(),p.end(),l)-p.begin();
    ll j = upper_bound(p.begin(),p.end(),r)-p.begin();
    j--;
    ll w = j-i+1;
    ll ans = A;
    if(w)
        ans = (r-l+1)*B*w;
    if(l==r || w==0) return ans;
    ll mid = (r+l)/2;
    ll res = dp(l,mid,A,B)+dp(mid+1,r,A,B);
    if(ans>res) return res;
    return ans;
}
ll solve(){
	cin>>n>>k>>A>>B;
    ll N = 1<<n;
    ll t;
    // for(int i=0;i<k;i++)p[i]=0;
    for(int i=0;i<k;i++){
        cin>>t;
        p.push_back(t);
    }
    // ll sufix[N+1];
    // for(int i=0;i<N;i++)
    //     sufix[i+1]=sufix[i]+p[i];
    // for(int i=0;i<k;i++)cout<<p[i]<<" ";cout<<endl;
    sort(p.begin(),p.end());
    // for(int i=0;i<k;i++)cout<<p[i]<<" ";cout<<endl;
    // return 1;
	return dp(1,N,A,B);
}

int main(){
	// T = nxt();
	while(T--)
        cout<<solve()<<endl;
	return 0;
}